src/Analyzer.Utilities.UnitTests/JsonLineNumberResolverTests.cs

```diff
@@ -0,0 +1,190 @@
+            // Resolve line number
+            var resolvedLineNumber = new JsonLineNumberResolver()
+                .ResolveLineNumberForOriginalTemplate(
+                    path,
+                    expandedTemplate,
+                    originalTemplate);
+
+            // Get expected line number
+            var tokenInOriginalTemplate = originalTemplate;
+            foreach (var pathSegment in pathInOrginalTemplate)
+            {
+                tokenInOriginalTemplate = tokenInOriginalTemplate[pathSegment];
+            }
+
+            var expectedLineNumber = (tokenInOriginalTemplate as IJsonLineInfo).LineNumber;
+            Assert.AreEqual(expectedLineNumber, resolvedLineNumber);
+        }
+
+        [DataTestMethod]
+        [DataRow("resources[4].type", DisplayName = "Extra resource")]
+        [DataRow("resources[3].type", DisplayName = "Extra copied resource with missing source copy loop")]
+        public void ResolveLineNumberForOriginalTemplate_UnableToFindEquivalentLocationInOriginal_Returns0(string path)
+        {
+            Assert.AreEqual(
+                0,
+                new JsonLineNumberResolver()
+                    .ResolveLineNumberForOriginalTemplate(
+                        path,
+                        expandedTemplate,
+                        originalTemplate));
+        }
+
+        [TestMethod]
+        [ExpectedException(typeof(ArgumentNullException))]
+        public void ResolveLineNumberForOriginalTemplate_PathIsNull_ThrowsException()
+        {
+            new JsonLineNumberResolver()
+                .ResolveLineNumberForOriginalTemplate(
+                    null,
+                    expandedTemplate,
+                    originalTemplate);
+        }
+
+        [TestMethod]
+        [ExpectedException(typeof(ArgumentNullException))]
+        public void ResolveLineNumberForOriginalTemplate_OriginalTemplateIsNull_ThrowsException()
+        {
+            new JsonLineNumberResolver()
+                .ResolveLineNumberForOriginalTemplate(
+                    "path",
+                    expandedTemplate,
+                    null);
+        }
+
+        [TestMethod]
+        [ExpectedException(typeof(ArgumentNullException))]
+        public void ResolveLineNumberForOriginalTemplate_ExpandedTemplateIsNullForScenario3_ThrowsException()
+        {
+            new JsonLineNumberResolver()
+                .ResolveLineNumberForOriginalTemplate(
+                    "resources[2]",
+                    null,
+                    originalTemplate);
+        }
+
+        [TestMethod]
+        public void ResolveLineNumberForOriginalTemplate_ExpandedTemplateIsNullForScenario1or2_ReturnsLineNumber()
+        {
+            new JsonLineNumberResolver()
+                .ResolveLineNumberForOriginalTemplate(
+                    "parameters.missingParameter",
+                    null,
+                    originalTemplate);
+        }
+
+        private JToken originalTemplate = JObject.Parse(
```

Comment:
Good call on readonly.  I put them down here because they are so long, and I wanted the actual tests to be readily accessible.  Do you still prefer they be at the top?

Reply:
I would say so. They can always be collapsed if needed. Reading the file top to bottom, they make more sense at the top.